### ISSUE TITLE
Non-visual access

### ISSUE THREAD

**Issue by Natalia-Hedlund** [osmandapp/OsmAnd#16758]
I would like to draw your attention to accessibility of Osmand for visually impaired users. Previously, there was an app, called Osmand Access, designed for partially sighted and blind people. The app is not supported for a long time. It works somehow on some devices, but it is not possible to install it from Playmarket anymore.

As a solution to this, it would be great if Igor @Poretsky, an experienced programmer, who developed Osmand Access app, could join the Osmandapp's team and improved the Accessibility plugin. At the moment it's functionality is very limited and does not give low-sighted users all possibilities for safe navigation.

**Comment by dmpr0** [comment]
Could you please provide more details about what exactly doesn't work with enabled Accessibility plugin?

**Comment by svtsvet** [comment]
App OsmAnd access doesn't install to my Pixel7. My phone isn't supporting this one. I installed OsmAnd from Play Market but i understand that it's impossible using this App without visual control. how Now there are only minimal functions in this plugin and it isn't providing navigation for the blind. May be Igor Poretsky could help us? can you write him? I know that many blind people use OsmAnd access but it works not so good now.

**Comment by svtsvet** [comment]
main features in OsmAnd accesss. Provided content description for almost all graphical interface elements (such as icons, buttons, etc.) that are essential for user interactions.
Unwanted events for accessibility service, leading to it’s inadequate behavior, are blocked.
Map response to the gestures is modified so that when map is bound to location the long tap on it brings up a context menu that corresponds to the anchor point of the map, i.e. the user location. It doesn’t depend on the touch point, which in the absence of visual control is poorly predictable. This special behaviour should be explicitly enabled in the program accessibility preferences if necessary.
Neighborhood exploration
This function is available via the context menu (choose item “Look around”). It displays a list of nearby objects located in the surroundings of the virtual location. The list may include not only the POI, but also favorites. You can configure the filter so that only the objects of your interest will be shown.
Voice control
A number of the program functions are available through voice control. This mode is called by shaking the device in the direction of its longitudinal axis, through the context menu or by pressing the button on the headset. When a headset is used it’s button can be used to interrupt speech input as well. The editor of the voice commands is available in the configuration of accessibility features of the program.
In the official version of OsmAnd application autoannounce (automatic announcement of the direction and distance to the destination) is issued only when the direction to the destination is changed. Here you can optionally specify a period of unconditional issuing of autoannounce. You can set minimal time interval between auto announcements in a range from 5 seconds to 5 minutes.
Accessibility preferences of the program related to the navigation process are set individually for each profile (car, bicycle, pedestrian). These include:
the prohibition of automatic re-route if you deviate from it, that is, if you are at a considerable distance from the specified route, or if the wrong direction of motion was detected;
setting approaching radius to the final and intermediate route points;
setting refresh rate of position.
All voice control and navigation functions remain available in background, when screen is off or locked.
Option that allows to adjust route deviation tolerance. This option can be set separately for each profile and it is taken in account only when accessibility mode is active. When it is set, user is warned about route failure during navigation as soon as distance from the route crosses specified limit.

**Comment by Tech4Blind** [comment]
Blind and visually impaired people cannot work with the visual objects on the map. To do this, the following functions must be additionally implemented in the Accessibility plugin:
The "Look around" mode, where objects are displayed as a list on request.
Working with voice commands with the ability to edit them
Shaking mode, where it is possible to assign voice control, the installation of user points or something else to shaking.
Setting up orienting hints – the ability to track an object selected by the user with an indication of the distance in a straight line and the direction to it.
The ability to switch modes of sound or tactile direction indications
Setting up route profiles with the ability to prohibit recalculation of route information, selection of radii to reach intermediate and waypoints and evasion from the trajectory of the route
Configuring the availability of route information.

Modified gestures do not work in the application – for example, double-tapping the Map object to get information and some other features.

Blind programmer Igor Poretsky created an alternative version of the application, where the above functionality was implemented in full. This application has been fully adapted to screen access programs.

**Comment by poretsky** [comment]
I am not a member of the team now, but I've noticed this issue because
it mentions me, though, somewhat late. Thus, now I'll try to clarify
the subject a bit.

All the extra features listed by @svtsvet in
[her comment](https://github.com/osmandapp/OsmAnd/issues/16758#issuecomment-1477658717)
as well as the ones mentioned in the
[next comment](https://github.com/osmandapp/OsmAnd/issues/16758#issuecomment-1477698746)
by @Tech4Blind were implemented in the course of OsmAnd Access project
that being a fork of OsmAnd in general was additionally furnished with
a bunch of accessibility related features. Some of them were then
implemented in OsmAnd as well when I was a member of its development team.

I agree that it is far from sufficient to make the app truly
accessible, it only makes an illusion of accessibility for those who
do not use it in that mode. On the other hand, OsmAnd Access was
accessible indeed, meanwhile, it is absolutely outdated for the
moment.

OsmAnd Access project has died some years ago and I see no reason to
reanimate it. It had some sense when accessibility means in Android
itself were in its early stage and it was necessary to invent some
tricks to overcome its deficiency. But modern Android is ready enough
to implement reasonable application accessibility in terms of the
universal design that I am convinced admirer. And, in principle, I can
do the work if there would be a request.

**Comment by vshcherb** [comment]
@poretsky can you please contact me by email victor.shcherb at osmand.net . I would like to double check what should be done once UI is created and what could be as extra steps / screens.

**Comment by vshcherb** [comment]
I think I could summarize it as
- Look around mode (that main topic of the issue)
Probably related with 
- The ability to switch modes of sound or tactile direction indications

Extra
- Overall screen accessibility (Check talkback functionality, proper icons description / etc
- Shaking mode, where it is possible to assign voice control
-----------------
Already Implemented (partially?):
- Repeat Voice instructions (implemented inside Voice Prompts -> Default option is manually by taping the widget)
- Setting up route profiles with the ability to prohibit recalculation of route information (it's possible to configure how to often to recalculate the route)
- Setting up orienting hints – the ability to track an object selected by the user with an indication of the distance in a straight line and the direction to it - some kind of markers could be used for that.

Lots of features of course but we need with the most important I think Look around mode and then later take step by step implementation.

**Comment by Tech4Blind** [comment]
Look around mode, modified gestures like double tap the map to get the current route instructions and the ability to track an object seem to be the most important features to work on.

**Comment by cyrmax** [comment]
Hello Community!
I hope that this issue is not already dead and some work will be done to make the Osmand accessibility better.

Unfortunately Osmand Access is not working on almost all modern devices. On some of them it refuses to install without extra actions because of too old and unsafe code and on other ones it just does not work at all because of old native libs which are of unsupported architecture.

So I hope that Osmand team and Igor Poretsky will do improve Osmand accessibility.

For me the following features are especially needed:
* Look around mode;
* Audible direction indication not only while walking against the route but also in search results, look around menu and other location lists (while some location is focused by screenreader the audible indication points to this location);
* Accessible map behavior (double tap with screenreader to get directions information or a distance depending on the current route and state; long press to open the location's context menu with "save", "directions" and more options);
* Announcing of the nearest map object when pressing "where I am" button;
* Displaying current street and address at the top of the screen (if detected);
* Ability to auto announce nearby PoI and favorites while moving;
* Audible indication to the destination point without a route (some blind users call it "point aproach mode");

Thanks in advance and best regards!

**Comment by a11cf0** [comment]
I totally agree with above comments. OsmAnd Access is dead and quickly becoming unusable on modern systems, but we don't have anything even remotely capable of replacing it. That's why having at least a working look around mode and stable direction indication in mainline OsmAnd would be awesome.

**Comment by grishakonzalaev** [comment]
I agree with all above comments. it needs to add look around function and set as destination point to check straight direction by sound or tactile compass.

**Comment by SergYekatzoon** [comment]
Hello dear community.
I am teaching blind users how to use a smartphone.
My listeners and I used OsmAnd-2.9.3-a11y+1 very actively.
However, after its last update, a lot of water has passed under the bridge.
I and more than 30 of my listeners dream of using OsmAnd on modern versions of Android.
The OsmAnd application has no analogues.
We would be very happy if accessibility features continued to be developed either in the main application or at least as a plugin.

With respect and gratitude to the development team
Sergey Ekaterinushkin

**Comment by masag689** [comment]
Hello dear OsmAnd development team!
I and more than 20 of my friends who do not speak English have previously used OsmAnd-2.9.3-a11y+1 very actively.
But this version is very outdated and practically does not work on modern versions of Android.
Please include Accessibility functions in the program, or make an appropriate addition.
Sincerely:
Sharov Sergey

**Comment by olegnikne** [comment]
Dear Osmand developers team,

I want to draw your attention to an important accessibility issue for blind and visually impaired people when using navigation applications. For over 12 years, I have been using GPS navigation, and for me, it is a crucial tool. Additionally, I am a computer and mobile literacy teacher, and my students are actively studying navigation applications, but they face accessibility issues for screen reader programs.

I am reaching out to you with a request to consider adapting the official Osmand application for people with vision impairments. It is important to implement additional features in the Accessibility plugin, such as the "Look around" mode, voice commands with editing capabilities, a shaking mode for voice control, and other convenient options.

I also want to acknowledge the work of Igor Poretsky, a blind programmer who created an alternative version of the application with full functionality for people with disabilities. This experience shows that adapting the application for blind users is possible and extremely important for creating an inclusive environment for all.

Thank you for your attention to this issue. I look forward to seeing steps towards improving the accessibility of the Osmand application for all users.

Sincerely,
Koposov Oleg

**Comment by cyrmax** [comment]
Hello! Recently I have purchased a new phone which doesn't support 32-bit applications at all.
So I had to switch from Osmand Access to an official app from Google Play.
I have noticed the following issues (they are not missing features but real accessibility bugs of already existing screens and options):
1. The screen with warning about speed control cameras which offers an ability to delete camera control notifications data is not accessible. Buttons are not shown to screenreaders and blind user cannot accept or decline this dialog.
2. When the Callman filter is on, audible and tactile indication of a direction works with big delays which does not allow blind user to use this feature reliably.
3. A top panel widget with a current street name sometimes shows just "street name widget" and does not report the current street to screenreaders.

Also I have a question about this issue in general:
do we need to report every bug and feature request here or should we create new issues for every found accessibility bug or to request new feature?

**Comment by DmitryAlexei** [comment]
This feature is planned for 4.9, so it's better to write your observations here so that they are considered.

**Comment by cyrmax** [comment]
@DmitryAlexei what feature are you talking about? Probably I didn't understand because English is not my native, sorry.

**Comment by DmitryAlexei** [comment]
@cyrmax I'm talking about "Look around mode" - it is still planned for 4.9

**Comment by Ishtef** [comment]
There are many alternative navigation apps for sighted users. OSMand is just one of many for them. For the blind users OSMand Access was one of the few full-fledged navigation applications. And if the features aren't ported to the main app, it's going to be a really big problem for the blind. Therefore, I fully support the integration of OSMand access functions into the main application.

**Comment by SavagingSky** [comment]
As someone who uses this app on per demand basis and is blind, I fully support what has been said here. Making the app accessible would benefit a lot of visually impared people.
